FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift

```diff
@@ -38,19 +38,73 @@ extension FetchedMessage {
 }
 
 // MARK: - MessageService
+enum MessageServiceError: Error {
+    case missedPassPhrase(_ rawMimeData: Data)
+    case wrongPassPhrase(_ rawMimeData: Data, _ passPhrase: String)
+    // Could not fetch keys
+    case emptyKeys
+}
+
 final class MessageService {
     private let messageProvider: MessageProvider
-    private let dataService: DataServiceType & KeyDataServiceType
+    private let keyService: KeyServiceType
+    private let passPhraseStorage: PassPhraseStorageType
     private let core: Core
 
     init(
         messageProvider: MessageProvider = MailProvider.shared.messageProvider,
-        dataService: DataServiceType & KeyDataServiceType = DataService.shared,
-        core: Core = Core.shared
+        keyService: KeyServiceType = KeyService(),
+        core: Core = Core.shared,
+        passPhraseStorage: PassPhraseStorageType = PassPhraseStorage(
+            storage: EncryptedStorage(),
+            emailProvider: DataService.shared
+        )
     ) {
         self.messageProvider = messageProvider
-        self.dataService = dataService
+        self.keyService = keyService
         self.core = core
+        self.passPhraseStorage = passPhraseStorage
+    }
+
+    func validateMessage(rawMimeData: Data, with passPhrase: String) -> Promise<FetchedMessage> {
+        Promise<FetchedMessage> { [weak self] resolve, reject in
+            guard let self = self else { return }
+
+            guard let keys = try? self.keyService.getPrivateKeys(with: passPhrase).get(), keys.isNotEmpty else {
+                return reject(MessageServiceError.emptyKeys)
+            }
+
+            // TODO: - Tom - is it possible to get longid of the key which was used for decryption?
+            let decrypted = try self.core.parseDecryptMsg(
+                encrypted: rawMimeData,
+                keys: keys,
+                msgPwd: nil,
+                isEmail: true
+            )
+
+            let isDecryptError = decrypted.blocks.isAnyError
+
+            if isDecryptError {
+                reject(MessageServiceError.wrongPassPhrase(rawMimeData, passPhrase))
+            } else {
+                keys
+                    .map { PassPhrase(value: passPhrase, longid: $0.longid) }
+                    .forEach { self.passPhraseStorage.savePassPhrase(with: $0, inStorage: false) }
+
+                let attachments = decrypted.blocks
+                    .filter(\.isAttachmentBlock)
+                    .map(MessageAttachment.init)
+
+                let fetchedMessage = FetchedMessage(
```

Comment:
Maybe `ProcessedMessage` would be better wording here? Then it would equally work for messages that were fetched without decrypting, and messages that were fetched and decrypted.

`FetchedMessage` name feels like the message is raw as received which is not true here.